src/read_pool.rs

```diff
@@ -0,0 +1,150 @@
+use futures::sync::oneshot;
+use futures::Future;
+use futures03::prelude::*;
+use kvproto::kvrpcpb::CommandPri;
+use std::future::Future as StdFuture;
+use tikv_util::future_pool::{self, FuturePool};
+use yatp::pool::{Local, Runner};
+use yatp::queue::Extras;
+use yatp::task::future::{Runner as FutureRunner, TaskCell};
+use yatp::Remote;
+
+use crate::storage::kv::{destroy_tls_engine, set_tls_engine, Engine};
+
+#[derive(Clone)]
+pub enum ReadPool {
+    FuturePools {
+        read_pool_high: FuturePool,
+        read_pool_normal: FuturePool,
+        read_pool_low: FuturePool,
+    },
+    Yatp(Remote<TaskCell>),
+}
+
+impl ReadPool {
+    pub fn spawn<F>(&self, f: F, priority: CommandPri, task_id: u64) -> Result<(), ReadPoolError>
+    where
+        F: StdFuture<Output = ()> + Send + 'static,
+    {
+        match self {
+            ReadPool::FuturePools {
+                read_pool_high,
+                read_pool_normal,
+                read_pool_low,
+            } => {
+                let pool = match priority {
+                    CommandPri::High => read_pool_high,
+                    CommandPri::Normal => read_pool_normal,
+                    CommandPri::Low => read_pool_low,
+                };
+
+                pool.spawn(move || Box::pin(f.never_error()).compat())?;
+            }
+            ReadPool::Yatp(remote) => {
+                let fixed_level = match priority {
+                    CommandPri::High => Some(0),
+                    CommandPri::Normal => None,
+                    CommandPri::Low => Some(2),
+                };
+                let extras = Extras::new_multilevel(task_id, fixed_level);
+                let task_cell = TaskCell::new(f, remote.clone(), extras);
+                remote.spawn(task_cell);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn spawn_handle<F, T, E>(
+        &self,
+        f: F,
+        priority: CommandPri,
+        task_id: u64,
+    ) -> Result<impl Future<Item = T, Error = E>, ReadPoolError>
+    where
+        F: StdFuture<Output = Result<T, E>> + Send + 'static,
+        T: Send + 'static,
+        E: Send + 'static,
+    {
+        let (tx, rx) = oneshot::channel::<Result<T, E>>();
+        self.spawn(
+            async move {
+                let _ = tx.send(f.await);
+            },
+            priority,
+            task_id,
+        )?;
+        Ok(rx.then(|res| res.expect("tx is dropped by the thread pool")))
```

Comment:
Why panic?

Reply:
The thread pool should not drop the future if it's not completed. Tokio threadpool's `spawn_handle` has the same behavior.

Reply:
That's not the same as yatp. Yatp will exit immediately when shutdown bit is detected and leave all the remaining tasks in channel.

Reply:
Ah, right. I thought wrong...